packages/core/src/controllers/publisher.ts

```diff
@@ -57,52 +61,67 @@ export class Publisher extends IPublisher {
-    const startPublish = Date.now();
-    let result;
-    let attempts = 1;
 
     try {
       /**
-       * Loop until the publish is successful or the timeout is reached
-       * The loop allows to retry to retry the publish in case of disconnect
+       * attempt to publish the payload for <initialPublishTimeout> seconds,
+       * if the publish fails, add the payload to the queue and it will be retried on every pulse
+       * until it is successfully published or <publishTimeout> seconds have passed
        */
-      while (result === undefined) {
-        // Terminate the publishing attempts if publisTimeout has been exceeded
-        if (Date.now() - startPublish > this.publishTimeout) {
-          throw new Error(failedPublishMessage);
-        }
-
-        this.logger.trace({ id, attempts }, `publisher.publish - attempt ${attempts}`);
-        const publish = await createExpiringPromise(
-          this.rpcPublish(topic, message, ttl, relay, prompt, tag, id, opts?.attestation).catch(
-            (e) => this.logger.warn(e),
-          ),
-          this.publishTimeout,
-          failedPublishMessage,
+      const publishPromise = new Promise(async (resolve) => {
+        const onPublish = ({ id }: { id: string }) => {
+          if (params.opts.id === id) {
+            this.removeRequestFromQueue(id);
+            this.relayer.events.removeListener(RELAYER_EVENTS.publish, onPublish);
+            resolve(params);
+          }
+        };
+        this.relayer.events.on(RELAYER_EVENTS.publish, onPublish);
+        const initialPublish = createExpiringPromise(
+          new Promise((resolve, reject) => {
+            this.rpcPublish({
+              topic,
+              message,
+              ttl,
+              prompt,
+              tag,
+              id,
+              attestation: opts?.attestation,
+            })
+              .catch((e) => {
+                this.logger.warn(e, e?.message);
+                reject(e);
+              })
+              .then(resolve);
+          }),
+          this.initialPublishTimeout,
+          `Failed initial publish, retrying.... id:${id} tag:${tag}`,
         );
-        result = await publish;
-        attempts++;
-
-        if (!result) {
-          // small delay before retrying so we can limit retry to max 1 time per second
-          // if network is down `rpcPublish` will throw immediately
-          await new Promise((resolve) => setTimeout(resolve, this.failedPublishTimeout));
-        }
-      }
-      this.relayer.events.emit(RELAYER_EVENTS.publish, params);
-      this.logger.debug(`Successfully Published Payload`);
+        await initialPublish
+          .then((result) => {
+            this.events.removeListener(RELAYER_EVENTS.publish, onPublish);
+            resolve(result);
+          })
+          .catch((e) => {
+            this.queue.set(id, { ...params, attempt: 1 });
+            this.logger.warn(e, e?.message);
+          });
```

Comment:
Not a big fan of how we're combining the use of the promise constructor with .then .catch and async/await + try/catch
I think the code could be easier to reason about if we were just using the promise constructor with async await and the try/catch block.